### PR TITLE
Add AMDGPU v0.4.x support

### PR DESCRIPTION
**Pull request by luraess**
- Implement kernel signal reuse
- Fix breaking changes introduced in AMDGPU 0.4.x API
- Temporally fix MPI `Const` naming (one needs to find a consistent solution in MPI.jl about `Consts` naming)
- Temporally no MPI compat in Project.toml as we need `#master` (to be v0.20.0 release)
- Need to find a strategy how to handle the `LocalPreferences.toml` file now needed and generated by MPIPreferences.jl

### CODE REVIEW

File: src/update_halo.jl [hunk 1]
```diff
@@ -709,8 +715,8 @@ function read_x2d!(gpurecvbuf::ROCDeviceArray{T}, A::ROCDeviceArray{T}, recvrang
 end
 
 # Write to the send buffer on the host from the array on the device (d2h).
-function write_d2h_async!(sendbuf::AbstractArray{T}, A::ROCArray{T}, sendranges::Array{UnitRange{T2},1}, signal::HSASignal) where T <: GGNumber where T2 <: Integer
-    locked_ptr = convert(Ptr{T}, AMDGPU.Mem.lock(pointer(sendbuf),sizeof(sendbuf),get_default_agent()))
+function write_d2h_async!(sendbuf::AbstractArray{T}, A::ROCArray{T}, sendranges::Array{UnitRange{T2},1}, signal::ROCSignal) where T <: GGNumber where T2 <: Integer
+    locked_ptr = convert(Ptr{T}, AMDGPU.Mem.lock(pointer(sendbuf),sizeof(sendbuf),AMDGPU.default_device()))
```
Comment (jpsamaroo) [inline src/update_halo.jl:711]: This should probably use `AMDGPU.device(sendbuf)`, if `sendbuf <: ROCArray`.

File: src/update_halo.jl [hunk 2]
```diff
@@ -723,8 +729,8 @@ function write_d2h_async!(sendbuf::AbstractArray{T}, A::ROCArray{T}, sendranges:
 end
 
 # Read from the receive buffer on the host and store on the array on the device (h2d).
-function read_h2d_async!(recvbuf::AbstractArray{T}, A::ROCArray{T}, recvranges::Array{UnitRange{T2},1}, signal::HSASignal) where T <: GGNumber where T2 <: Integer
-    locked_ptr = convert(Ptr{T}, AMDGPU.Mem.lock(pointer(recvbuf),sizeof(recvbuf),get_default_agent()))
+function read_h2d_async!(recvbuf::AbstractArray{T}, A::ROCArray{T}, recvranges::Array{UnitRange{T2},1}, signal::ROCSignal) where T <: GGNumber where T2 <: Integer
+    locked_ptr = convert(Ptr{T}, AMDGPU.Mem.lock(pointer(recvbuf),sizeof(recvbuf),AMDGPU.default_device()))
```
Comment (jpsamaroo) [inline src/update_halo.jl:725]: Same as above

**Review by jpsamaroo — APPROVED** [review]
Seems good to me!

File: src/update_halo.jl (hunk 1 above)
Comment (luraess) [reply src/update_halo.jl:711]: ~~@jpsamaroo, @omlins how should one implement the `sendbuf <: ROCArray` condition in Julian's suggestion to replace `AMDGPU.default_device()` by `AMDGPU.device(sendbuf)`?~~ EDIT: see 👇

File: src/update_halo.jl (hunk 2 above)
Comment (luraess) [reply src/update_halo.jl:733]: @omlins same as above

File: src/update_halo.jl (hunk 1 above)
Comment (luraess) [reply src/update_halo.jl:711]: @jpsamaroo, `sendbuf` (and `recvbuf`) are actually no `ROCArray` and thus `AMDGPU.device(sendbuf)` fails. Either we leave it as such, or we may rather use `AMDGPU.device(A)` which is the corresponding `ROCArray` we want to copy to/from.

Comment (jpsamaroo) [reply src/update_halo.jl:711]: Oh, sorry, I totally misread this! You want this to be host-locked, so you'd need `first(AMDGPU.devices(:cpu))`, probably.

Comment (luraess) [reply src/update_halo.jl:711]: For host-locking, HSA will actually return a new pointer from the given device, thus I guess it should still be either `get_default_agent()` or `AMDGPU.device(A)`.

```
help?> AMDGPU.Mem.lock
  lock(ptr::Ptr, bytesize::Integer, device::ROCDevice)
  lock(ptr, bytesize)
  lock(a::Array, device)
  lock(a)

  Page-lock a host pointer allocated by the OS allocator and return a new pointer from the given device. For more information, see hsa_amd_memory_lock().

  See also: unlock

julia> 
```

Comment (luraess) [reply src/update_halo.jl:711]: Any thoughts on ☝️ @jpsamaroo ?

**Comment by luraess** [comment]
Closing in favour of #55

**Pull request closed without merging**